File: exercism/ruby/linked-list/linked_list.rb
class Deque
  def initialize
    @head = nil
    @tail = nil
  end

  def push(data)
    node = Node.new(data)

    case @head
    when nil
      @head = node
      @tail = node
    when @tail
      @head = node
      @head.next = @tail
      @tail.prev = @head
    else
      node.next = @head
      @head.prev = node
      @head = node
    end

    self
  end

  def pop
    return nil if @head.nil?

    node = @head

    case @head
    when @tail
      @head = nil
      @tail = nil
    else
      @head = @head.next
      @head.prev = nil
    end

    node.data
  end

  def unshift(data)
    node = Node.new(data)

    case @tail
    when nil
      @tail = node
      @head = node
    when @head
      @tail = node
      @head.next = @tail
      @tail.prev = @head
    else
      node.prev = @tail
      @tail.next = node
      @tail = node
    end

    self
  end

  def shift
    return nil if @tail.nil?

    node = @tail

    case @tail
    when @head
      @head = nil
      @tail = nil
    else
      @tail = @tail.prev
      @tail.next = nil
    end

    node.data
  end

  private

  class Node

    attr_reader :data
    attr_accessor :prev, :next

    def initialize(data = nil)
      @data = data
      @next = nil
      @prev = nil
    end
  end

  private_constant :Node
end

File: exercism/javascript/linked-list/linked-list.js
function Node(data) {
  this.data = data;
  this.next = null;
  this.prev = null;
}

Node.prototype.append = function append(node) {
  this.next = node;
  node.prev = this;
};

Node.prototype.prepend = function prepend(node) {
  this.prev = node;
  node.next = this;
};

function LinkedList() {
  this.first = null;
  this.last = null;
}

Object.defineProperty(LinkedList.prototype, "empty", {
  get: function() { return this.last === null; }
});

LinkedList.prototype.push = function push(data) {
  var node = new Node(data);

  if (this.empty) {
    this.last = node;
    this.first = node;
  } else {
    this.last.append(node);
    this.last = node;
  }
};

LinkedList.prototype.pop = function pop() {
  var data = this.last.data;
  this.last = this.last.prev;
  this.last ? this.last.next = null : this.first = null;
  return data;
};

LinkedList.prototype.unshift = function unshift(data) {
  var node = new Node(data);
  if (this.empty) {
    this.first = node;
    this.last = node;
  } else {
    this.first.prepend(node);
    this.first = node;
  }
};

LinkedList.prototype.shift = function shift() {
  var data = this.first.data;
  this.first = this.first.next;
  this.first ? this.first.prev = null : this.last = null;
  return data;
};

LinkedList.prototype.count = function count() {
  var node = this.first;
  var counter = 0;
  while (node) {
    counter += 1;
    node = node.next;
  }
  return counter;
};

// delete *all* occurences of data from linked list
LinkedList.prototype.delete = function remove(data) {
  var node = this.first;

  while (node) {
    if (node.data === data) {
      if (node.prev && node.next) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
      } else if (node.prev) {
        node.prev.next = null;
        this.last = node.prev;
      } else if (node.next) {
        node.next.prev = null;
        this.first = node.next;
      } else {
        this.first = null;
        this.last = null;
      }
    }
    node = node.next;
  }
};

module.exports = LinkedList;
